community/rfcs/24-07-11-OPEA-Agent.md

```diff
@@ -0,0 +1,278 @@
+  llm_model_id: str
+  recursion_limit: int
+  tools: file_path or dict
+  
+  # Tools definition
+  [tool_name]:
+    description: str
+    callable_api: choices([http://xxxx, xxx.py:func_name])
+    env: str
+    pip_dependencies: str # sep by ,
+    args_schema:
+      query:
+        type: choices([int, str, bool])
+        description: str
+    return_output: str
+  ```
+
+  #### Agent Role microservice definition - 'Planner': 
+  Agent without tools. Planner only contains LLM endpoints as planner, certain strategies to complete an optimized plan. 
+
+  configuration:
+  ```
+  strategy: choices([react, planexec, humanInLoopPlanExec])
+  require_human_feedback: bool
+  llm_endpoint_url: str
+  llm_engine: choices([tgi, vllm, openai])
+  llm_model_id: str
+  recursion_limit: int
+  require_human_feedback: bool
+  ```
+
+  #### Agent Role microservice definition - 'Executor': 
+  Tools executors. Executor is used to process input with registered tools.
+
+  Configuration:
+  ```
+  [tool_name]:
+  description: str
+  callable_api: choices([http://xxxx, xxx.py:func_name])
+  env: str
+  pip_dependencies: str # sep by ,
+  args_schema:
+    query:
+      type: choices([int, str, bool])
+      description: str
+  return_output: str
+  ```
+
+  > Any microservcice follow this spec can be registered as role in Part3-graph-based
+  
+### Part3. 'Multi Agent' system overview
+
+We planned to provide multi-agent system in two phases.
+
+* Phase I: Hierarchical Multi Agents
+  1.	In this design, only top-layer Agent will be exposed to OPEA mega flow. And only ‘Agent’ microservice will be used to compose Hierarchical Multi Agents system.
+  2.	Users are only allowed to use yaml files to provide tools configuration, high-level instructions text and hierarchical relationship between agents.
+  3.	This design simplifies the agent configuration, using simple yaml definition can still be used to compose a multi agent system to handle complex tasks.
+  > Detailed configuration please refer to Part3.1
+  ![image](https://github.com/user-attachments/assets/be3bef3a-a1c9-4059-a8a1-e8e52e0d6c16)
+
+
+* Phase II: Graph-Based Multi Agent
+  1.	In this design, we provide user a new SDK to compose a graph-based multi agents system with conditional edge to define all strategic rules. 
+  2.	Enterprise user will be able to use python code to wrap either ‘agent’, ‘planner’ or tools  as ‘Role’ and add conditional edges between them for complex task agent design. 
+  3.	This design provides user enough flexibility to handle very complex tasks and also provide flexibility to handle resource management when certain tools are running way slower than others.
+  > Detailed configuration please refer to Part3.2
+  ![image](https://github.com/user-attachments/assets/35b36f64-eaa1-4f05-b25e-b8bea013680d)
+
+#### Part3.1 Hierarchical Multi Agents
+__Example 1__: ‘Single Agent megaservice’
+Only 1 agent is presented in this configuration. 
```

Comment:
Is this Agent configuration written by developer or auto-generated?

Reply:
The 'opea_micro_service' and 'opea_mega_service' configuration should be prepared by enterprise devops team. For docker-based side, ip and port should be provided. For GMC/k8s side, we should leverage a python script to obtain domain_name:port from k8s.

The 'tool' yaml should be generated by enterprise developer according to their company tools/available APIs, and then end_user should be able to cherry-pick some tools from the tool store.